File: apps/pos-android/core-domain/src/test/kotlin/com/snappos/domain/MoneyTest.kt
package com.snappos.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * These are the same cases as `packages/contracts/src/money.test.ts`, on
 * purpose.
 *
 * Two implementations of one rule set drift. That is not a risk, it is a
 * certainty given enough time, and in a pricing engine drift means a customer
 * was charged something the reports disagree with. Mirroring the assertions is
 * the cheap half of the defence; the language neutral conformance fixtures that
 * both engines run are the other half.
 *
 * If a case is changed here, change it there in the same commit.
 */
class MoneyTest {

  @Test
  fun `minor units round trip through major-unit strings`() {
    assertEquals("24.99", Money.fromMajor("24.99").toMajorString())
    assertEquals("24.90", Money.fromMajor("24.9").toMajorString())
    assertEquals("24.00", Money.fromMajor("24").toMajorString())
    assertEquals("0.05", Money.fromMajor("0.05").toMajorString())
    assertEquals("-3.50", Money.fromMajor("-3.50").toMajorString())
    assertEquals(2499L, Money.fromMajor("24.99").minor)
  }

  @Test
  fun `more precision than a cent is refused, not truncated`() {
    assertThrows(MoneyFormatException::class.java) { Money.fromMajor("24.999") }
    assertThrows(MoneyFormatException::class.java) { Money.deserialize("24.99") }
  }

  @Test
  fun `the classic float failure does not occur`() {
    // 0.1 + 0.2 != 0.3 in IEEE-754. This is the whole reason the type exists.
    assertNotEquals(0.3, 0.1 + 0.2, 0.0)
    assertEquals(Money.fromMajor("0.30"), Money.fromMajor("0.10") + Money.fromMajor("0.20"))

    // A hundred nickels is exactly five dollars, every time.
    val nickels = List(100) { Money.fromMajor("0.05") }
    assertEquals("5.00", Money.sum(nickels).toMajorString())
  }

  @Test
  fun `a realistic cart totals to the cent`() {
    val lines = listOf(
      Money.fromMajor("24.99") * 2,
      Money.fromMajor("6.49"),
      Money.fromMajor("3.99"),
    )
    val subtotal = Money.sum(lines)
    assertEquals("60.46", subtotal.toMajorString())

    val taxable = subtotal + Money.fromMajor("-6.49")
    val tax = taxable.applyRate("0.0825")
    assertEquals("4.45", tax.toMajorString())
    assertEquals("58.42", (taxable + tax).toMajorString())
  }

  @Test
  fun `tax rounds half up, away from zero`() {
    // 1000 * 0.0825 = 82.5 exactly. Half up gives 83, banker's gives 82.
    assertEquals(83L, Money(1000).applyRate("0.0825").minor)
    assertEquals(-83L, Money(-1000).applyRate("0.0825").minor)
    assertEquals(1L, Money(1).applyRate("0.5").minor)
    assertEquals(0L, Money(1).applyRate("0.49").minor)
  }

  @Test
  fun `tax on a rate binary floating point cannot represent`() {
    // 8.25 * 100 in floating point is 824.9999999999999.
    val amount = Money.fromMajor("8.25")
    assertEquals(825L, amount.minor)
    assertEquals(68L, amount.applyRate("0.0825").minor)
  }

  @Test
  fun `allocate never loses or invents a cent`() {
    val parts = Money.fromMajor("10.00").allocate(3)
    assertEquals(listOf("3.34", "3.33", "3.33"), parts.map { it.toMajorString() })
    assertEquals(Money.fromMajor("10.00"), Money.sum(parts))

    for ((amount, n) in listOf("0.01" to 3, "99.99" to 7, "-10.00" to 3, "0.00" to 4)) {
      val split = Money.fromMajor(amount).allocate(n)
      assertEquals(n, split.size)
      assertEquals("$amount across $n", Money.fromMajor(amount), Money.sum(split))
    }
  }

  @Test
  fun `weighted allocation sums back exactly`() {
    // A $5.00 cart discount across lines of 49.98, 6.49 and 3.99.
    val weights = listOf(
      Money.fromMajor("49.98"),
      Money.fromMajor("6.49"),
      Money.fromMajor("3.99"),
    )
    val shares = Money.fromMajor("-5.00").allocateByWeight(weights)
    assertEquals(Money.fromMajor("-5.00"), Money.sum(shares))
    assertEquals(3, shares.size)
    // The largest line absorbs the most discount.
    assertTrue(shares[0] < shares[1])
  }

  @Test
  fun `weights that are all zero fall back to an even split`() {
    val shares = Money.fromMajor("1.00")
      .allocateByWeight(listOf(Money.ZERO, Money.ZERO, Money.ZERO))
    assertEquals(Money.fromMajor("1.00"), Money.sum(shares))
  }

  @Test
  fun `money crosses the wire as a string`() {
    val value = Money.fromMajor("24.99")
    assertEquals("2499", value.serialize())
    assertEquals(value, Money.deserialize("2499"))

    // Above 2^53 a JSON number would silently lose precision. A Long does not.
    val huge = Money.deserialize("9007199254740993")
    assertEquals(huge, Money.deserialize(huge.serialize()))
    assertNotEquals(huge.serialize(), huge.minor.toDouble().toLong().toString())
  }

  @Test
  fun `cost keeps six decimals until it is posted`() {
    // A case of 12 at $5.00 is $0.416667 a unit. Rounding to cents first and
    // multiplying back gives 4.92, and margin reports drift from there.
    assertEquals(500L, costToMinor("0.416667", 12).minor)
    assertEquals("5.00", costToMinor("0.416667", 12).toMajorString())
  }

  @Test
  fun `subtraction and negative balances behave`() {
    assertEquals("-2.50", (Money.fromMajor("5.00") - Money.fromMajor("7.50")).toMajorString())
    assertEquals(Money.ZERO, Money.sum(emptyList()))
  }

  @Test
  fun `overflow throws rather than wrapping`() {
    // A silently wrapped total is the worst possible failure for money.
    assertThrows(ArithmeticException::class.java) {
      Money(Long.MAX_VALUE) + Money(1)
    }
  }
}

File: packages/contracts/src/money.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
  money,
  fromMajor,
  toMajorString,
  add,
  subtract,
  sum,
  multiply,
  applyRate,
  allocate,
  allocateByWeight,
  serialize,
  deserialize,
  costToMinor,
  decimal,
  MoneyError,
  ZERO,
} from './money.js';

test('minor units round trip through major-unit strings', () => {
  assert.equal(toMajorString(fromMajor('24.99')), '24.99');
  assert.equal(toMajorString(fromMajor('24.9')), '24.90');
  assert.equal(toMajorString(fromMajor('24')), '24.00');
  assert.equal(toMajorString(fromMajor('0.05')), '0.05');
  assert.equal(toMajorString(fromMajor('-3.50')), '-3.50');
  assert.equal(toMajorString(money('-9223372036854775808')), '-92233720368547758.08');
  assert.equal(fromMajor('24.99'), 2499n);
});

test('dollars passed where cents were expected is refused, not rounded', () => {
  // 24.99 as minor units would be a 100x error. The only safe answer is to stop.
  assert.throws(() => money(24.99), MoneyError);
  assert.throws(() => fromMajor('24.999'), MoneyError);
  assert.throws(() => money('24.99'), MoneyError);
});

test('the classic float failure does not occur', () => {
  // 0.1 + 0.2 !== 0.3 in IEEE-754. This is the whole reason the type exists.
  assert.notEqual(0.1 + 0.2, 0.3);
  assert.equal(add(fromMajor('0.10'), fromMajor('0.20')), fromMajor('0.30'));

  // A hundred nickels is exactly five dollars, every time.
  const nickels = Array.from({ length: 100 }, () => fromMajor('0.05'));
  assert.equal(toMajorString(sum(nickels)), '5.00');
});

test('a realistic cart totals to the cent', () => {
  const lines = [multiply(fromMajor('24.99'), 2), fromMajor('6.49'), fromMajor('3.99')];
  const subtotal = sum(lines);
  assert.equal(toMajorString(subtotal), '60.46');

  const promo = fromMajor('-6.49');
  const taxable = add(subtotal, promo);
  const tax = applyRate(taxable, '0.0825');
  assert.equal(toMajorString(tax), '4.45');
  assert.equal(toMajorString(add(taxable, tax)), '58.42');
});

test('tax rounds half up, away from zero', () => {
  // 1000 * 0.0825 = 82.5 exactly. Half up gives 83, banker's rounding gives 82.
  // Two identical receipts must never disagree, so half up it is.
  assert.equal(applyRate(money(1000n), '0.0825'), 83n);
  assert.equal(applyRate(money(-1000n), '0.0825'), -83n);
  assert.equal(applyRate(money(1n), '0.5'), 1n);
  assert.equal(applyRate(money(1n), '0.49'), 0n);
});

test('tax on a rate that binary floating point cannot represent', () => {
  // 0.0825 has no exact float. 8.25 * 100 in JS is 824.9999999999999.
  const amount = fromMajor('8.25');
  assert.equal(amount, 825n);
  assert.equal(applyRate(amount, '0.0825'), 68n);
});

test('allocate never loses or invents a cent', () => {
  const parts = allocate(fromMajor('10.00'), 3);
  assert.deepEqual(parts.map(toMajorString), ['3.34', '3.33', '3.33']);
  assert.equal(sum(parts), fromMajor('10.00'));

  for (const [amount, n] of [
    ['0.01', 3],
    ['99.99', 7],
    ['-10.00', 3],
    ['0.00', 4],
  ] as const) {
    const split = allocate(fromMajor(amount), n);
    assert.equal(split.length, n);
    assert.equal(sum(split), fromMajor(amount), `${amount} across ${n}`);
  }
});

test('weighted allocation sums back exactly, remainder to the largest line', () => {
  // A $5.00 cart discount across lines of 49.98, 6.49 and 3.99.
  const weights = [fromMajor('49.98'), fromMajor('6.49'), fromMajor('3.99')];
  const shares = allocateByWeight(fromMajor('-5.00'), weights);
  assert.equal(sum(shares), fromMajor('-5.00'));
  assert.equal(shares.length, 3);
  // The largest line absorbs the most discount.
  assert.ok(shares[0]! < shares[1]!);
});

test('weights that are all zero fall back to an even split', () => {
  const shares = allocateByWeight(fromMajor('1.00'), [ZERO, ZERO, ZERO]);
  assert.equal(sum(shares), fromMajor('1.00'));
});

test('money crosses the wire as a string, never a number', () => {
  const value = fromMajor('24.99');
  assert.equal(serialize(value), '2499');
  assert.equal(deserialize('2499'), value);

  // JSON.stringify throws on a bigint, which is the behaviour we want: it makes
  // an accidental float round trip impossible rather than merely discouraged.
  assert.throws(() => JSON.stringify({ total: value }), TypeError);

  // Above Number.MAX_SAFE_INTEGER a number would silently lose precision.
  const huge = money('9007199254740993');
  assert.equal(deserialize(serialize(huge)), huge);
  assert.notEqual(Number(serialize(huge)).toString(), serialize(huge));
});

test('cost keeps six decimals until it is posted', () => {
  // A case of 12 at $5.00 is $0.416667 a unit. Rounding that to cents and
  // multiplying back gives 4.92, not 5.00, and margin reports drift from there.
  const unitCost = decimal('0.416667');
  assert.equal(costToMinor(unitCost, 12), 500n);
  assert.equal(toMajorString(costToMinor(unitCost, 12)), '5.00');

  assert.throws(() => decimal('0.4166667'), MoneyError);
});

test('subtraction and negative balances behave', () => {
  assert.equal(toMajorString(subtract(fromMajor('5.00'), fromMajor('7.50'))), '-2.50');
  assert.equal(sum([]), ZERO);
});

test('values outside the Postgres bigint range are rejected', () => {
  assert.throws(() => money('9223372036854775808'), MoneyError);
  assert.throws(() => multiply(money(2n), 999999999999999999999n as unknown as bigint), MoneyError);
});
